### PR TITLE
disk: sdhc: Switch to clock frequency from DTS

### LINKED ISSUE #28092: Make SPI speed of SDHC card configurable
The SD-card-spi driver has a hardcoded 400kHz and 4Mhz SPI bus speed, it would be good to make this configurable (maybe via DT?)

### LINKED ISSUE #32996: SPI speed when using SDHC via SPI in Zephyr
I'm using the Nordic nRF52840-DK board to breadboard an SDHC card interface in Zephyr using SPI.    I have it doing "basic" operations, but am trying to reconcile what I'm physically seeing versus what is documented and what I see in the code.   I was hoping to run the SPI at 8 MHz, so I included a statement in the .overlay file:   clock-frequency = <8000000>;  in the SPI specification.   I also put the statement:   spi-max-frequency = <8000000>;  in the specification of the SDcard sub-device (under the SPI specification).   The Zephyr SDHC documentation page seems to imply that the spi-max-frequency parameter is the SPI running frequency applied after initialization.   But it appears that hardcoded into disk_access_spi_sdhc.c  are statements that assign the speed to be 400 KHz during initialization, then 4 MHz for later operations.    But then looking at the waveform captured by a logic analyzer, it appears that the SCK signal is actually running at around 250 KHz.  What am I missing?   Thanks!

### PR DESCRIPTION
**Pull request by richardbarlow**
The "spi-max-frequency" property already exists, but was unused. This
now sets the SPI clock frequency to this value (limited to 24MHz) once
initialisation is complete.

Due to the nature of the SPI API, it is necessary to have two separate
configuration structures to switch clock speed as some SPI drivers only
compare pointers to detect changes.

Fixes: #32996, Fixes: #28092, #22906

Improves issue reported in #26513, but does not implement CMD6 to
switch to High Speed mode.

Signed-off-by: Rich Barlow <rich@bennellick.com>

### CODE REVIEW

**Comment by richardbarlow** [comment]
This supersedes #26319. It rebases the change to master and incorporates the requested architectural changes.

File: drivers/disk/sdmmc_spi.c [hunk 1]
```diff
@@ -18,9 +18,11 @@ LOG_MODULE_REGISTER(sdmmc_spi, CONFIG_SDMMC_LOG_LEVEL);
 #include "sdmmc_sdhc.h"
```
Comment (jfischer-no) [inline drivers/disk/sdmmc_spi.c:18]: "Fixes: #32996" is sufficient.

_"...Improves issue reported in #26513, but does not implement CMD6 to
switch to High Speed mode."_ does not belong to the commit message, please remove.

File: drivers/disk/sdmmc_spi.c [hunk 2]
```diff
@@ -979,7 +992,26 @@ static int disk_spi_sdhc_init(const struct device *dev)
...
+	.init_cfg = {
+		.frequency = SDHC_SPI_INIT_SPEED,
+		.operation = SPI_WORD_SET(8) | SPI_HOLD_ON_CS,
+		.slave = DT_REG_ADDR(SPI_SDHC_NODE),
+#if DT_SPI_DEV_HAS_CS_GPIOS(SPI_SDHC_NODE)
+		.cs = &sdhc_spi_data_0.cs,
+#endif
+	},
+	.oper_cfg = {
+		.frequency = MIN(SDHC_SPI_MAX_OPER_SPEED, DT_INST_PROP(0, spi_max_frequency)),
```
Comment (jfischer-no) [inline drivers/disk/sdmmc_spi.c:1005]: Please reshape,

```diff
diff --git a/drivers/disk/sdmmc_spi.c b/drivers/disk/sdmmc_spi.c
index 0b79b777b4..3afa60b287 100644
--- a/drivers/disk/sdmmc_spi.c
+++ b/drivers/disk/sdmmc_spi.c
@@ -1002,7 +1002,8 @@ static const struct sdhc_spi_config sdhc_spi_cfg_0 = {
 #endif
        },
        .oper_cfg = {
-               .frequency = MIN(SDHC_SPI_MAX_OPER_SPEED, DT_INST_PROP(0, spi_max_frequency)),
+               .frequency = MIN(SDHC_SPI_MAX_OPER_SPEED,
+                                DT_INST_PROP(0, spi_max_frequency)),
                .operation = SPI_WORD_SET(8) | SPI_HOLD_ON_CS,
                .slave = DT_REG_ADDR(SPI_SDHC_NODE),
 #if DT_SPI_DEV_HAS_CS_GPIOS(SPI_SDHC_NODE)
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Thanks, just few comments.

File: drivers/disk/sdmmc_spi.c (hunk 1 above)
Comment (jfischer-no) [reply drivers/disk/sdmmc_spi.c:18]: >Due to the nature of the SPI API, it is necessary to have two separate
configuration structures to switch clock speed as some SPI drivers only
compare pointers to detect changes.

This is correct behavior that every driver should follow.

**Review by jfischer-no — APPROVED** [review]

File: drivers/disk/sdmmc_spi.c [hunk 3]
```diff
@@ -979,7 +990,26 @@ static int disk_spi_sdhc_init(const struct device *dev)
 
 static struct sdhc_spi_data sdhc_spi_data_0;
 
+static const struct sdhc_spi_config sdhc_spi_cfg_0 = {
+	.init_cfg = {
+		.frequency = SDHC_SPI_INIT_SPEED,
+		.operation = SPI_WORD_SET(8) | SPI_HOLD_ON_CS,
+		.slave = DT_REG_ADDR(SPI_SDHC_NODE),
+#if DT_SPI_DEV_HAS_CS_GPIOS(SPI_SDHC_NODE)
+		.cs = &sdhc_spi_data_0.cs,
```
[the remark this answers is not in the document]
Comment (anangl) [reply drivers/disk/sdmmc_spi.c:1001]: > I'm not 100% happy with this (and the matching line in the `oper_cfg` initialisation). The configs are const, but require a pointer to the `struct spi_cs_control`, which has to live somewhere. Currently this lives in the `sdhc_spi_data_0` struct, which means we must directly reference it here.

You can move this structure to `sdhc_spi_cfg_0`, as all its members can be initialized at compile time (in particular, `gpio_dev` to `DEVICE_DT_GET(DT_SPI_DEV_CS_GPIOS_CTLR(SPI_SDHC_NODE))`).

File: drivers/disk/sdmmc_spi.c [hunk 4]
```diff
@@ -41,6 +43,24 @@ struct sdhc_spi_data {
 #endif
 };
 
+struct sdhc_spi_config {
+	const struct spi_config init_cfg;
+	const struct spi_config oper_cfg;
```
Comment (anangl) [inline drivers/disk/sdmmc_spi.c:48]: These `const`s here are redundant, as the whole structure is instantiated as `const`.

File: drivers/disk/sdmmc_spi.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (richardbarlow) [reply drivers/disk/sdmmc_spi.c]: Thanks for the pointer. I've just updated the PR based on your suggestion and I'm now much happier with it.

File: drivers/disk/sdmmc_spi.c (hunk 4 above)
Comment (richardbarlow) [reply drivers/disk/sdmmc_spi.c]: Removed

**Review by anangl — APPROVED** [review]
